### PR TITLE
vim-patch:8.1.0864 Make 'scrolloff' and 'sidescrolloff' options window local

### PR DESCRIPTION
**Pull request by sophec**
As noted in #11006, recent versions of vim allow scrolloff to be a window local-global option. This is a feature I was sorely missing (as I like to turn off `scrolloff` for `:terminal`s, but not for every other window). This patch changes these two options to be window global-local. This is my first neovim PR, so any tips and feedback are welcome.

The following patch is included: [8.1.0864: brammool: Cannot have a local value for 'scrolloff' and 'sidescrolloff'.](https://github.com/vim/vim/commit/375e3390078e740d3c83b0c118c50d9a920036c7?diff=split)

### CODE REVIEW

**Comment by bfredl** [comment]
By experience #11445, we should turn global options into global-local ones, and not into pure local ones. This reduces the breakage of existing configs and expectations. 

> This is not a vim PR, however. 

It is still useful to tag the commits with the vim-patch thing, to mark the patches as included, the functionality is what matters, not the exact code. Though I do suspect looking at the vim patches might save some time here.

**Comment by WillEccles** [comment]
@bfredl I believe this is actually global-local, given my .vimrc's `set so=3` still works just the same for every window, but every window can set it independently using `setl`. This is how I understand global-local to function. Am I incorrect? "Window local" in the PR title is incorrect, I will admit.

**Comment by WillEccles** [comment]
I have also update the docs for this, since I had set them to be "local to window."

**Comment by bfredl** [comment]
That it happens to work when testing one (1) specific scenario doesn't mean it is correct. Try open a few windows and then change and reload your vimrc, it will not work as a global-local option.

**Comment by WillEccles** [comment]
I see now. I was unaware that a global-local option worked this way. I will revert my changes and attempt to implement Moolenar's changes from the original patch. Shall I make a new PR for this?

**Comment by bfredl** [comment]
no, just push to this PR.

**Comment by WillEccles** [comment]
Can do.

**Comment by WillEccles** [comment]
Did not mean to close this, simply meant to revert changes. Will reopen with new changes.

**Comment by WillEccles** [comment]
Just applied the patch. At the moment, the following behavior occurs:

1. Open nvim and do `:set`. You will see your `scrolloff` as specified in `.vimrc`.
2. Open a new window within nvim. This window will have no `scrolloff` set.

At the moment I must take a break from debugging this, but if anyone has any ideas, I'd appreciate any input.

**Comment by justinmk** [comment]
CI also shows a failing test

```
[  FAILED  ] test/functional/options/num_options_spec.lua @ 44: :set validation setlocal and setglobal validate values
test/functional/options/num_options_spec.lua:9: Expected objects to be the same.
Passed in:
(string) 'E487'
Expected:
(string) 'E49'

stack traceback:
	test/functional/options/num_options_spec.lua:9: in function 'should_fail'
	test/functional/options/num_options_spec.lua:68: in function <test/functional/options/num_options_spec.lua:44>
```


You can [run the test](https://github.com/neovim/neovim/blob/master/test/README.md#running-tests) like this:

    TEST_FILE=test/functional/options/num_options_spec.lua make functionaltest

**Comment by WillEccles** [comment]
I did not get around to running tests yesterday because I was out of time. Will resume debugging today.

**Comment by WillEccles** [comment]
It's just occurred to me that buffer global-local might make more sense than window global-local. Anyone have any thoughts? I should be able to swap this out pretty quickly if anyone else agrees.

**Comment by bfredl** [comment]
No, it should be a window option. It controls how a buffer is displayed in a window.

**Comment by WillEccles** [comment]
Regarding tests: patch 8.1.0864 sets the error for invalid `scrolloff` to be the same as `sidescrolloff`, so I will update the test to accommodate. The tests fail because (of course) the error resulting from an invalid `scrolloff` is not what it used to be.

**Comment by WillEccles** [comment]
Actually, I would appreciate some assistance. I am looking at `num_options_spec.lua`. It's failing because in 8.1.0864, -1 is a valid local value (which resets it to global), but it's not a valid global value. This test only works for *both* global *and* local. Is there somewhere else that these should be tested? Do I remove the tests?

**Comment by WillEccles** [comment]
For clarity, these are the problem lines: https://github.com/neovim/neovim/blob/master/test/functional/options/num_options_spec.lua#L68-L69

**Comment by bfredl** [comment]
Probably the simplest to write an own block for `scrolloff` + `sidescrolloff`, just like for `window` option.

**Comment by WillEccles** [comment]
I have added tests for setting the global value to -1 (which should fail) and setting the local value to -1 (which should pass). These tests pass on my machine.

**Comment by WillEccles** [comment]
No idea what to do about the failing CI—it appears to my eye to be unrelated to my changes?

**Comment by bfredl** [comment]
Please make sure some commit contains the string " vim-patch:8.2.0235: cannot have a local value for 'scrolloff' and 'sidescrolloff'", the exact formulation (up to including the second comma, I think) is needed for our infra to autodetect merged patches.

File: src/nvim/option.c [hunk 1]
```diff
@@ -7475,3 +7494,22 @@ dict_T *get_winbuf_options(const int bufopt)
 
   return d;
 }
+
+/*
```
Comment (bfredl) [inline src/nvim/option.c]: use /// comments before functions

File: src/nvim/search.c [hunk 2]
```diff
@@ -2276,8 +2278,8 @@ showmatch(
       State = SHOWMATCH;
       ui_cursor_shape();                /* may show different cursor shape */
       curwin->w_cursor = mpos;          /* move to matching char */
-      p_so = 0;                         /* don't use 'scrolloff' here */
-      p_siso = 0;                       /* don't use 'sidescrolloff' here */
+      *so = 0;                         /* don't use 'scrolloff' here */
```
Comment (bfredl) [inline src/nvim/search.c]: change /* comments both of the added lines and the neighbours to // style

**Comment by bfredl** [comment]
Mostly linter failures. Try reading the "lint" job output and travis and fix them (mostly `/* */` comments)

**Comment by WillEccles** [comment]
Sorry about the comments—was just reading the vim patch. Will fix.

**Comment by WillEccles** [comment]
> Please make sure some commit message contains the string " vim-patch:8.2.0235: cannot have a local value for 'scrolloff' and 'sidescrolloff'",

Can this be in the "description" of a commit? (i.e. `git commit -m "did thing" -m "vim-patch...."`)

**Comment by WillEccles** [comment]
Never mind, followed this, which I missed the first time around: https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim#pull-requests

**Comment by bfredl** [comment]
it can be on any line in the message, which is useful, so that a commit can contain multiple vim patches.

File: src/nvim/option.c [hunk 3]
```diff
@@ -900,11 +900,18 @@ set_option_default(
       if (options[opt_idx].indir == PV_SCROLL) {
         win_comp_scroll(curwin);
       } else {
-        *(long *)varp = (long)(intptr_t)options[opt_idx].def_val[dvi];
+        long def_val = (long)options[opt_idx].def_val[dvi];
+        if ((long *)varp == &curwin->w_p_so
+                || (long *)varp == &curwin->w_p_siso)
+            // 'scrolloff' and 'sidescrolloff' local values have a
```
Comment (bfredl) [inline src/nvim/option.c]: indentation (should be same as next line)

File: src/nvim/option.c [hunk 4]
```diff
@@ -900,11 +900,18 @@ set_option_default(
       if (options[opt_idx].indir == PV_SCROLL) {
         win_comp_scroll(curwin);
       } else {
-        *(long *)varp = (long)(intptr_t)options[opt_idx].def_val[dvi];
+        long def_val = (long)options[opt_idx].def_val[dvi];
+        if ((long *)varp == &curwin->w_p_so
+                || (long *)varp == &curwin->w_p_siso)
```
Comment (bfredl) [inline src/nvim/option.c]: `if` should always use braces (the linter will tell you about the relevant cases)

File: src/nvim/option.c (hunk 3 above)
Comment (WillEccles) [reply src/nvim/option.c]: Referring to line 905?

Comment (bfredl) [reply src/nvim/option.c]: Referring to the line at which I commented. Though the issue is actually that TAB:s are used on the subsequent lines, replace it all with 2-space indent.

Comment (WillEccles) [reply src/nvim/option.c]: Interesting, I didn't realize they were tabs. I don't use tabs. Will check later.

**Comment by WillEccles** [comment]
Every time I fix all the linter's issues and commit it finds more and more issues to fix.

**Comment by WillEccles** [comment]
Not sure what's up with the CI. Looks like the macOS build for travis failed because of a missing command.

**Comment by justinmk** [comment]
> Every time I fix all the linter's issues and commit it finds more and more issues to fix.

That is intentional, otherwise we would need to reformat the entire codebase at once (not a bad idea, but meanwhile...) see https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#automated-builds-ci . You can run lint check locally:

    make lint



> Not sure what's up with the CI. Looks like the macOS build for travis failed because of a missing command.

some travis problem. restarted.

**Comment by WillEccles** [comment]
Not sure what to do about the failing CI, any ideas/advice would be appreciated.

**Comment by b0o** [comment]
Pinging this - seems like @WillEccles could use some help with CI. This would be a great feature to have.

**Comment by KillTheMule** [comment]
Travis failure is unrelated. Bsd build fails on a test I somehow remember to be flaky, but I'm not real sure. I suggest you push a dummy commit to retrigger CI so we can see if the problem persists.

**Comment by WillEccles** [comment]
I'll push a commit which fixes some of the merge conflicts. We will see if the CI is happy then.

**Comment by WillEccles** [comment]
Looks like Travis passed this time, only the strange BSD one left to finish. I have looked at the BSD one that failed and could not determine why it wasn't working. I don't think it had to do with any of my code, but I suppose I could be wrong?

**Comment by KillTheMule** [comment]
I don't see any failures anymore, so it's fine.

ping @justinmk

File: src/nvim/move.c [hunk 5]
```diff
@@ -1349,8 +1361,8 @@ void scroll_cursor_top(int min_scroll, int always)
   linenr_T old_topline = curwin->w_topline;
   linenr_T old_topfill = curwin->w_topfill;
   linenr_T new_topline;
-  assert(p_so <= INT_MAX);
-  int off = (int)p_so;
+  int off = (int)get_scrolloff_value();
+  assert(off <= INT_MAX);
```
Comment (bfredl) [inline src/nvim/move.c]: This assert now doesn't do anything. But probably it should just be deleted, range checks are done when options are _set_, and need not be duplicated to every point an option is _used_.

File: src/nvim/move.c [hunk 6]
```diff
@@ -1724,9 +1737,9 @@ void cursor_correct(void)
    * How many lines we would like to have above/below the cursor depends on
    * whether the first/last line of the file is on screen.
    */
-  assert(p_so <= INT_MAX);
-  int above_wanted = (int)p_so;
-  int below_wanted = (int)p_so;
+  int above_wanted = (int)get_scrolloff_value();
+  int below_wanted = (int)get_scrolloff_value();
+  assert(below_wanted <= INT_MAX);
```
Comment (bfredl) [inline src/nvim/move.c]: same.

**Comment by WillEccles** [comment]
Removed those assertions.

File: src/nvim/option.c [hunk 7]
```diff
@@ -7485,3 +7505,18 @@ dict_T *get_winbuf_options(const int bufopt)
 
   return d;
 }
+
+/// Return the effective 'scrolloff' value for the current window, using the
+/// global value when appropriate.
+long get_scrolloff_value(void)
+{
+    return curwin->w_p_so < 0 ? p_so : curwin->w_p_so;
```
Comment (bfredl) [inline src/nvim/option.c]: 2 space indent.

File: src/nvim/option.c [hunk 8]
```diff
@@ -7485,3 +7505,18 @@ dict_T *get_winbuf_options(const int bufopt)
...
+}
+
+/// Return the effective 'sidescrolloff' value for the current window, using the
+/// global value when appropriate.
+long get_sidescrolloff_value(void)
+{
+    return curwin->w_p_siso < 0 ? p_siso : curwin->w_p_siso;
```
Comment (bfredl) [inline src/nvim/option.c]: ditto.

**Comment by bfredl** [comment]
LGTM otherwise. could squash also (but make sure to keep the ` vim-patch:` strings in the message).

**Comment by WillEccles** [comment]
Fixed. Squashing is not something I have ever mastered, but whoever merges can use the squash and merge option and set the commit message themselves.

**Pull request merged**